src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs

```diff
@@ -0,0 +1,201 @@
+
+                    try
+                    {
+                        await _consumer.ResetAsync(taskInfo.TaskId, new TaskResultData(TaskResult.Fail, ex.Message), taskInfo.RunId, MaxRetryCount, cancellationToken);
+                    }
+                    catch (Exception resetEx)
+                    {
+                        _logger.LogError(resetEx, "Task {0} failed to reset.", taskInfo.TaskId);
+                    }
+
+                    // Not complete the task for retriable exception.
+                    return;
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogError(ex, "Task {0} failed.", taskInfo.TaskId);
+                    result = new TaskResultData(TaskResult.Fail, ex.Message);
+                }
+
+                try
+                {
+                    await _consumer.CompleteAsync(taskInfo.TaskId, result, task.RunId, cancellationToken);
+                    _logger.LogInformation("Task {0} completed.", taskInfo.TaskId);
+                }
+                catch (Exception completeEx)
+                {
+                    _logger.LogError(completeEx, "Task {0} failed to complete.", taskInfo.TaskId);
+                }
+            }
+            finally
+            {
+                _activeTaskRecordsForKeepAlive.Remove(taskInfo.TaskId, out _);
+            }
+        }
+
+        private async Task KeepAliveTasksAsync(CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Start to keep alive task message.");
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                Task intervalDelayTask = Task.Delay(TaskHeartbeatIntervalInSeconds);
+                KeyValuePair<string, ITask>[] activeTaskRecords = _activeTaskRecordsForKeepAlive.ToArray();
+
+                foreach ((string taskId, ITask task) in activeTaskRecords)
+                {
+                    try
+                    {
+                        if (task.IsCancelling())
+                        {
+                            continue;
+                        }
+
+                        bool shouldCancel = false;
+                        try
+                        {
+                            TaskInfo taskInfo = await _consumer.KeepAliveAsync(taskId, task.RunId, cancellationToken);
+                            shouldCancel |= taskInfo.IsCanceled;
+                        }
+                        catch (TaskNotExistException notExistEx)
+                        {
+                            _logger.LogError(notExistEx, "Task {0} not exist or runid not match.", taskId);
+                            shouldCancel = true;
+                        }
+
+                        if (shouldCancel)
+                        {
+                            task.Cancel();
+                        }
+                    }
+                    catch (Exception ex)
```

Comment:
Can we catch a specific exception instead of being generic? i.e. `TaskCancelledException`

Reply:
We catch general exception here is to prevent any possible exception throw here to break the keep alive thread. Specific exception would be handled inside.